File: Aditya Vardhan Singh/Other Problems/Day 13/148 Sort List/C++/148 Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode * mergeList(ListNode * left, ListNode * right) {
        ListNode * dummy_head = new ListNode(-1);
        ListNode * temp = dummy_head;

        while (left && right) {
            if (left->val < right->val) {
                temp->next = left;
                left = left->next;
            } else {
                temp->next = right;
                right = right->next;
            }
            temp = temp->next;
        }

        if (left) {
            temp->next = left;
        } else {
            temp->next = right;
        }

        return dummy_head->next;
    }
    ListNode * getMid(ListNode * head) {
        if (!head || !head->next) {
            return head;
        }

        ListNode * slow = head;
        ListNode * fast = head->next;

        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        return slow;
    }
    ListNode * sortList(ListNode* head) {
        if (!head || !head->next) { // already sorted
            return head;
        }

        ListNode * mid = getMid(head);
        ListNode * left = head;
        ListNode * right = mid->next;
        mid->next = nullptr;

        left = sortList(left);
        right = sortList(right);

        return mergeList(left, right);
    }
};

File: Aditya Vardhan Singh/Other Problems/Day 13/148 Sort List/Python/148 Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def merge_list(self, left, right):
        dummy_head = ListNode(-1)
        temp = dummy_head

        while left and right:
            if left.val < right.val:
                temp.next = left
                left = left.next
            else:
                temp.next = right
                right = right.next
            temp = temp.next
        
        if left:
            temp.next = left
        else:
            temp.next = right

        return dummy_head.next

    def find_middle_of_list(self, head):
        if head is None or head.next is None:
            return head
        
        slow = head
        fast = head.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        return slow

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        
        mid = Solution.find_middle_of_list(self, head)
        left = head
        right = mid.next
        mid.next = None
        left = Solution.sortList(self, left)
        right = Solution.sortList(self, right)
        return Solution.merge_list(self, left, right)
    